### DISCUSSION TITLE
Best way to implement versioning?

### DISCUSSION THREAD

**Discussion by SR84** [microsoft/vs-streamjsonrpc#984] [Q&A] [answered]
This is a great project that I really enjoy using.

Now, I face the problem that I need to adapt my interface, and it will lead to breaking changes. What would be the best way to implement a kind of versioning, similar to the versioning in WebApi projects?

SR

**Comment by AArnott** [comment, marked as answer]
I don't know how WebApi projects do their versioning. But at a high level, I'd say you have two strategies to consider:

1. Avoid breaking changes by adding new methods without altering or removing existing ones. Note that in .NET adding members to an `interface` is a breaking change, but across JSON-RPC, it is _not_ a breaking change to add methods to an interface. 
2. Fork your service, and expose it at two endpoints. One endpoint exposes the old API and the second endpoint exposes the new one. The endpoints might be at two unique URLs, or they could be at the same URL but differentiate by HTTP header. Or if URL doesn't apply, you could change the port number on the new endpoint. 
The most advanced form of this could be keep just one endpoint, but add a handshake RPC method that newer clients will call first. In this method's implementation, you'll replace the v1 'local RPC target' object with your v2 object. In essence, you'll be providing your clients a way to 'upgrade' from v1 to v2 in the middle of their RPC session.

**Reply by SR84** [reply]
Thank you for your feedback. As you described, I am now forking my services and creating a separate channel for each new version using MultiplexingStream. To determine the server version, there is now a channel called "Neutral" where I query the server version. Now, the client has all the necessary information.

Great thing, the MultiplexingStream class from Nerdbank.Streams. It seems to also be from your hand. :-)
